### PR TITLE
feat(test-optimization): create final_status tag on test event for cypress

### PR DESCRIPTION
**Pull request by cbasitodx**
What does this PR do?
Create a @test.final_status tag that is added to every cypress test event that represents the final execution of that test (there can be multiple executions of a single test due to retry mechanisms being active)

Motivation
Tracers send each test execution as a span to the backend. When retry mechanisms are enabled a single test can run several times with different outcomes. These mechanisms differ in behavior. Some stop once the test reaches the desired state such as ATR ending when a test passes. Others always run a fixed number of attempts such as EFD or Attempt to fix.

To ensure the correct final result for the customer some intermediate outcomes are suppressed and the testing framework receives an adjusted final status. For example if an early EFD attempt passes but later attempts fail the test is still reported as passed and the failing attempts are hidden to avoid failing CI pipelines.

Both external and internal users have requested a way to query tests in Datadog by this final status. Their main goal is to build monitors and alerts for hard failures on default branches.

To support this a new tag will be added to the last retry of a test named: @test.final_status:pass|fail|skip

### CODE REVIEW

File: integration-tests/cypress/cypress.spec.js [hunk 1]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+            assert.ok(afterHookFailed, 'test with failing after() not found')
+            assert.strictEqual(afterHookFailed.meta[TEST_FINAL_STATUS], 'fail')
+
+            // Hooks: before() failure causes tests in the suite to be skipped
+            const beforeHookSkipped = tests.find(t =>
+              t.resource === 'cypress/e2e/hook-describe-error.cy.js.before will be skipped'
+            )
+            assert.ok(beforeHookSkipped, 'test skipped by before() not found')
+            assert.strictEqual(beforeHookSkipped.meta[TEST_FINAL_STATUS], 'skip')
+          }, 25000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness
```suggestion
          }, 60000)
```

File: integration-tests/cypress/cypress.spec.js [hunk 2]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+            ))
+            assert.strictEqual(newTestsWithHooks.length, NUM_RETRIES_EFD + 1)
+            newTestsWithHooks.forEach((test, idx) => {
+              if (idx < newTestsWithHooks.length - 1) {
+                assert.ok(!(TEST_FINAL_STATUS in test.meta))
+              } else {
+                assert.strictEqual(test.meta[TEST_FINAL_STATUS], 'pass')
+              }
+            })
+          }, 30000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness
```suggestion
          }, 60000)
```

File: integration-tests/cypress/cypress.spec.js [hunk 3]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+            assert.strictEqual(disabledTest.meta[TEST_STATUS], 'skip')
+            assert.strictEqual(disabledTest.meta[TEST_MANAGEMENT_IS_DISABLED], 'true')
+            assert.strictEqual(disabledTest.meta[TEST_FINAL_STATUS], 'skip')
+
+            const disabledWithHooks = tests.find(t => t.meta[TEST_NAME] === 'disabled with hooks is disabled')
+            assert.ok(disabledWithHooks)
+            assert.strictEqual(disabledWithHooks.meta[TEST_STATUS], 'skip')
+            assert.strictEqual(disabledWithHooks.meta[TEST_MANAGEMENT_IS_DISABLED], 'true')
+            assert.strictEqual(disabledWithHooks.meta[TEST_FINAL_STATUS], 'skip')
+          }, 25000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness 
```suggestion
          }, 60000)
```

File: integration-tests/cypress/cypress.spec.js [hunk 4]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+            assert.strictEqual(quarantinedWithHooks.meta[TEST_FINAL_STATUS], 'skip')
+
+            // Non-quarantined test with hooks: final_status='pass'
+            const passingWithHooks = tests.find(t =>
+              t.meta[TEST_NAME] === 'quarantined with hooks passes normally'
+            )
+            assert.ok(passingWithHooks)
+            assert.strictEqual(passingWithHooks.meta[TEST_STATUS], 'pass')
+            assert.strictEqual(passingWithHooks.meta[TEST_FINAL_STATUS], 'pass')
+          }, 25000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness
```suggestion
          }, 60000)
```

File: integration-tests/cypress/cypress.spec.js [hunk 5]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+
+            // Test body passes but afterEach throws; failure suppressed because test is quarantined
+            const quarantinedTest = tests.find(t =>
+              t.meta[TEST_NAME] === 'quarantined with failing afterEach is quarantined'
+            )
+            assert.ok(quarantinedTest)
+            assert.strictEqual(quarantinedTest.meta[TEST_STATUS], 'fail')
+            assert.strictEqual(quarantinedTest.meta[TEST_MANAGEMENT_IS_QUARANTINED], 'true')
+            assert.strictEqual(quarantinedTest.meta[TEST_FINAL_STATUS], 'skip')
+          }, 25000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness
```suggestion
          }, 60000)
```

File: integration-tests/cypress/cypress.spec.js [hunk 6]
```diff
@@ -4376,5 +4377,588 @@ moduleTypes.forEach(({
...
+            sorted.forEach((test, idx) => {
+              if (idx < sorted.length - 1) {
+                assert.ok(!(TEST_FINAL_STATUS in test.meta),
+                  `TEST_FINAL_STATUS should not be set on intermediate ATF run ${idx}`)
+              } else {
+                // All attempts failed → hasPassedAllRetries=false → final_status='fail'
+                assert.strictEqual(test.meta[TEST_FINAL_STATUS], 'fail')
+              }
+            })
+          }, 25000)
```
Comment (juan-fernandez) [inline integration-tests/cypress/cypress.spec.js]: let's set this to 60000 as the others to minimize flakiness
```suggestion
          }, 60000)
```

**Review by juan-fernandez — COMMENTED** [review]
great job! Really thorough tests. I've found a rare edge case. I'll push the fix in the branch so it's faster

**Comment by juan-fernandez** [comment]
I've pushed the fix and the addition of the integration test in https://github.com/DataDog/dd-trace-js/pull/8025/changes/98eaafee90aa21a66f5fbdef377010d7abd7d9f2.

**Review by juan-fernandez — APPROVED** [review]

**Review by gnufede — APPROVED** [review]
